k2/csrc/ragged.cu

```diff
@@ -395,43 +415,37 @@ void RaggedShape::Check() {
 
 RaggedShape RaggedShape2(Array1<int32_t> *row_splits, Array1<int32_t> *row_ids,
                          int32_t cached_tot_size) {
-  if (!row_splits && !row_ids) {
-    K2_LOG(FATAL) << "At least one of row_splits and row_ids must be defined";
-  }
+  K2_CHECK(row_splits != nullptr || row_ids != nullptr)
+      << "At least one of row_splits and row_ids must be defined";
+  ContextPtr ctx = ::GetContext(row_splits, row_ids);
   if (cached_tot_size != -1) {
     if (row_ids != nullptr) K2_CHECK_EQ(cached_tot_size, row_ids->Dim());
-    if (row_splits != nullptr) {  // caution: next check may be slow...
-      const auto &row_splits_ref = *row_splits;
-      K2_CHECK_EQ(cached_tot_size, row_splits_ref[row_splits->Dim() - 1]);
+    if (row_splits != nullptr) {
+      // may be slow as it may copy memory from device to host
+      K2_CHECK_EQ(cached_tot_size,
+                  row_splits->operator[](row_splits->Dim() - 1));
```

Comment:
The code would be much simpler if there was a `back()` method for `Array1<T>`.

Reply:
Added, thanks